File: backend/app/services/planetary_analytics.py
from __future__ import annotations

from collections import defaultdict
from datetime import datetime, timedelta, timezone
from typing import Any

from sqlalchemy import func, select
from sqlalchemy.orm import Session, selectinload

from app.models import EveType, PlanetaryColony, PlanetaryPin, PlanetaryRoute
from app.models.planetary_analytics import PlanetaryProductionSnapshot
from app.services.planetary_industry import (
    DEFAULT_DECAY_FACTOR,
    DEFAULT_NOISE_FACTOR,
    extractor_dogma_factors,
    extractor_program_projection,
)
from app.services.planetary_analytics_engine import evaluate_planetary_analytics_with_engine

# Stable SDE inventory group identifiers for PI products.
PI_TIER_BY_GROUP_ID = {
    1033: "P0",
    1042: "P1",
    1034: "P2",
    1040: "P3",
    1041: "P4",
}
PI_TIER_LABELS = {
    "P0": "Raw resources",
    "P1": "Basic commodities",
    "P2": "Refined commodities",
    "P3": "Specialized commodities",
    "P4": "Advanced commodities",
}
FACTORY_CYCLE_SECONDS = {"P1": 1800, "P2": 3600, "P3": 3600, "P4": 3600}


def commodity_tier(eve_type: EveType | None) -> str | None:
    return PI_TIER_BY_GROUP_ID.get(eve_type.group_id) if eve_type else None


def _aware(value: datetime | None) -> datetime | None:
    if value is None or value.tzinfo is not None:
        return value
    return value.replace(tzinfo=timezone.utc)


def _elapsed_days(start: datetime | None, end: datetime) -> float:
    start = _aware(start)
    end = _aware(end) or end
    if start is None or start >= end:
        return 0
    return min((end - start).total_seconds() / 86400, 30)


def _previous_rows(
    db: Session,
    character_id: int,
) -> tuple[datetime | None, dict[tuple[int, int], PlanetaryProductionSnapshot]]:
    captured_at = db.scalar(
        select(func.max(PlanetaryProductionSnapshot.captured_at)).where(
            PlanetaryProductionSnapshot.character_id == character_id
        )
    )
    if captured_at is None:
        return None, {}
    rows = db.scalars(
        select(PlanetaryProductionSnapshot).where(
            PlanetaryProductionSnapshot.character_id == character_id,
            PlanetaryProductionSnapshot.captured_at == captured_at,
        )
    ).all()
    return captured_at, {(row.pin_id, row.product_type_id): row for row in rows}


def record_planetary_production_snapshot(
    db: Session,
    character_id: int,
    captured_at: datetime | None = None,
) -> int:
    captured_at = _aware(captured_at) or datetime.now(timezone.utc)
    previous_at, previous = _previous_rows(db, character_id)
    colonies = db.scalars(
        select(PlanetaryColony)
        .where(PlanetaryColony.character_id == character_id)
        .options(
            selectinload(PlanetaryColony.pins),
            selectinload(PlanetaryColony.routes).selectinload(PlanetaryRoute.content_type),
        )
    ).all()
    extractor_type_ids = {
        pin.type_id
        for colony in colonies
        for pin in colony.pins
        if pin.extractor_cycle_time is not None
    }
    dogma = extractor_dogma_factors(db, extractor_type_ids)
    type_ids = {
        int(pin.extractor_product_type_id)
        for colony in colonies
        for pin in colony.pins
        if pin.extractor_product_type_id
    } | {
        int(route.content_type_id)
        for colony in colonies
        for route in colony.routes
    }
    type_rows = {
        row.type_id: row
        for row in db.scalars(select(EveType).where(EveType.type_id.in_(type_ids))).all()
    }
    created = 0

    for colony in colonies:
        inbound = {route.destination_pin_id for route in colony.routes}
        outbound: dict[int, dict[int, float]] = defaultdict(lambda: defaultdict(float))
        for route in colony.routes:
            outbound[route.source_pin_id][route.content_type_id] += float(route.quantity or 0)

        for pin in colony.pins:
            if pin.extractor_product_type_id and pin.extractor_cycle_time:
                product = type_rows.get(pin.extractor_product_type_id)
                tier = commodity_tier(product)
                if tier != "P0":
                    continue
                decay, noise, _ = dogma.get(
                    pin.type_id,
                    (DEFAULT_DECAY_FACTOR, DEFAULT_NOISE_FACTOR, "documented_default"),
                )
                projection = extractor_program_projection(
                    install_time=pin.install_time,
                    expiry_time=pin.expiry_time,
                    cycle_time=pin.extractor_cycle_time,
                    quantity_per_cycle=pin.extractor_qty_per_cycle,
                    decay_factor=decay,
                    noise_factor=noise,
                    now=captured_at,
                )
                rate = float(projection["average_daily_output"])
                remaining = float(projection["remaining_output"])
                if _aware(pin.expiry_time) and _aware(pin.expiry_time) <= captured_at:
                    rate = 0
                prior = previous.get((pin.pin_id, pin.extractor_product_type_id))
                produced = 0.0
                if (
                    prior
                    and _aware(prior.program_started_at) == _aware(pin.install_time)
                    and prior.projected_remaining_units is not None
                ):
                    produced = max(float(prior.projected_remaining_units) - remaining, 0)
                db.add(
                    _snapshot(
                        character_id,
                        captured_at,
                        previous_at,
                        colony,
                        pin,
                        product,
                        tier,
                        "extractor",
                        rate,
                        remaining,
                        produced,
                        pin.install_time,
                    )
                )
                created += 1

            if pin.schematic_id is None or pin.pin_id not in inbound:
                continue
            for product_type_id, quantity_per_cycle in outbound.get(pin.pin_id, {}).items():
                product = type_rows.get(product_type_id)
                tier = commodity_tier(product)
                cycle_seconds = FACTORY_CYCLE_SECONDS.get(tier or "")
                if not product or not tier or tier == "P0" or not cycle_seconds:
                    continue
                rate = quantity_per_cycle * 86400 / cycle_seconds
                prior = previous.get((pin.pin_id, product_type_id))
                produced = 0.0
                if prior:
                    produced = min(float(prior.projected_units_per_day), rate) * _elapsed_days(
                        previous_at, captured_at
                    )
                db.add(
                    _snapshot(
                        character_id,
                        captured_at,
                        previous_at,
                        colony,
                        pin,
                        product,
                        tier,
                        "factory",
                        rate,
                        None,
                        produced,
                        None,
                    )
                )
                created += 1
    db.flush()
    return created


def _snapshot(
    character_id: int,
    captured_at: datetime,
    previous_at: datetime | None,
    colony: PlanetaryColony,
    pin: PlanetaryPin,
    product: EveType | None,
    tier: str,
    source_kind: str,
    rate: float,
    remaining: float | None,
    produced: float,
    program_started_at: datetime | None,
) -> PlanetaryProductionSnapshot:
    return PlanetaryProductionSnapshot(
        character_id=character_id,
        captured_at=captured_at,
        interval_started_at=previous_at,
        planet_id=colony.planet_id,
        solar_system_id=colony.solar_system_id,
        pin_id=pin.pin_id,
        source_kind=source_kind,
        product_type_id=product.type_id if product else 0,
        commodity_tier=tier,
        unit_volume=float(product.volume or 0) if product else 0,
        projected_units_per_day=max(rate, 0),
        projected_remaining_units=remaining,
        estimated_units_since_previous=max(produced, 0),
        program_started_at=program_started_at,
    )


def planetary_analytics_summary(
    db: Session,
    days: int,
    character_ids: set[int],
    anonymous_character_ids: set[int] | None = None,
) -> dict[str, Any]:
    cutoff = datetime.now(timezone.utc) - timedelta(days=days)
    history = []
    all_rows = []
    if character_ids:
        history = db.scalars(
            select(PlanetaryProductionSnapshot)
            .where(
                PlanetaryProductionSnapshot.character_id.in_(character_ids),
                PlanetaryProductionSnapshot.captured_at >= cutoff,
            )
            .options(
                selectinload(PlanetaryProductionSnapshot.character),
                selectinload(PlanetaryProductionSnapshot.product_type),
            )
            .order_by(PlanetaryProductionSnapshot.captured_at)
        ).all()
        all_rows = db.scalars(
            select(PlanetaryProductionSnapshot)
            .where(PlanetaryProductionSnapshot.character_id.in_(character_ids))
            .options(
                selectinload(PlanetaryProductionSnapshot.character),
                selectinload(PlanetaryProductionSnapshot.product_type),
            )
            .order_by(
                PlanetaryProductionSnapshot.character_id,
                PlanetaryProductionSnapshot.captured_at.desc(),
            )
        ).all()
    latest_at: dict[int, datetime] = {}
    current: list[PlanetaryProductionSnapshot] = []
    for row in all_rows:
        latest = latest_at.setdefault(row.character_id, row.captured_at)
        if row.captured_at == latest:
            current.append(row)

    anonymous_character_ids = anonymous_character_ids or set()
    payload = {
        "schema_version": "eqm.planetary-analytics.v1",
        "days": days,
        "cutoff": cutoff.isoformat(),
        "history": [_snapshot_engine_row(row) for row in history],
        "current": [_snapshot_engine_row(row) for row in current],
        "anonymous_character_ids": sorted(anonymous_character_ids),
    }
    return evaluate_planetary_analytics_with_engine(
        payload=payload,
        python_result=lambda: _planetary_analytics_summary_python(
            days,
            cutoff,
            history,
            current,
            anonymous_character_ids,
        ),
    )


def _snapshot_engine_row(row: PlanetaryProductionSnapshot) -> dict[str, Any]:
    return {
        "character_id": row.character_id,
        "character_name": row.character.name,
        "product_type_id": row.product_type_id,
        "product_name": row.product_type.name,
        "tier": row.commodity_tier,
        "unit_volume": float(row.unit_volume or 0),
        "estimated_units_since_previous": float(row.estimated_units_since_previous or 0),
        "projected_units_per_day": float(row.projected_units_per_day or 0),
        "interval_started_at": _aware(row.interval_started_at).isoformat() if row.interval_started_at else None,
        "captured_at": (_aware(row.captured_at) or row.captured_at).isoformat(),
    }


def _planetary_analytics_summary_python(
    days: int,
    cutoff: datetime,
    history: list[PlanetaryProductionSnapshot],
    current: list[PlanetaryProductionSnapshot],
    anonymous_character_ids: set[int],
) -> dict[str, Any]:
    if not history and not current:
        return _empty_summary(days)

    aggregates: dict[tuple[int, int], dict[str, Any]] = {}
    for row in history:
        key = (row.character_id, row.product_type_id)
        item = aggregates.setdefault(key, _aggregate_seed(row))
        units = _windowed_estimate(row, cutoff)
        item["estimated_units"] += units
        item["estimated_volume"] += units * row.unit_volume
    for row in current:
        key = (row.character_id, row.product_type_id)
        item = aggregates.setdefault(key, _aggregate_seed(row))
        item["current_units_per_day"] += row.projected_units_per_day
        item["current_volume_per_day"] += row.projected_units_per_day * row.unit_volume

    character_products = sorted(
        aggregates.values(),
        key=lambda item: (item["estimated_volume"], item["current_volume_per_day"]),
        reverse=True,
    )
    products: dict[int, dict[str, Any]] = {}
    tiers = {
        tier: {
            "tier": tier,
            "label": PI_TIER_LABELS[tier],
            "estimated_units": 0.0,
            "estimated_volume": 0.0,
            "current_units_per_day": 0.0,
            "current_volume_per_day": 0.0,
            "products": set(),
            "characters": set(),
        }
        for tier in PI_TIER_LABELS
    }
    for item in character_products:
        product = products.setdefault(
            item["product_type_id"],
            {
                "product_type_id": item["product_type_id"],
                "product_name": item["product_name"],
                "tier": item["tier"],
                "estimated_units": 0.0,
                "estimated_volume": 0.0,
                "current_units_per_day": 0.0,
                "current_volume_per_day": 0.0,
                "top_character": None,
                "_top_score": -1.0,
            },
        )
        for field in (
            "estimated_units",
            "estimated_volume",
            "current_units_per_day",
            "current_volume_per_day",
        ):
            product[field] += item[field]
            tiers[item["tier"]][field] += item[field]
        score = item["estimated_units"] or item["current_units_per_day"]
        if item["character_id"] not in anonymous_character_ids and score > product["_top_score"]:
            product["_top_score"] = score
            product["top_character"] = item["character_name"]
        tiers[item["tier"]]["products"].add(item["product_type_id"])
        tiers[item["tier"]]["characters"].add(item["character_id"])

    product_rows = []
    for item in products.values():
        item.pop("_top_score", None)
        product_rows.append(item)
    product_rows.sort(
        key=lambda item: (item["estimated_volume"], item["current_volume_per_day"]),
        reverse=True,
    )
    tier_rows = []
    for item in tiers.values():
        item["product_count"] = len(item.pop("products"))
        item["character_count"] = len(item.pop("characters"))
        tier_rows.append(item)
    return {
        "days": days,
        "has_history": any(item["estimated_units"] > 0 for item in character_products),
        "cards": {
            "estimated_volume": sum(item["estimated_volume"] for item in character_products),
            "current_volume_per_day": sum(
                item["current_volume_per_day"] for item in character_products
            ),
            "product_count": len(products),
            "character_count": len(
                {item["character_id"] for item in character_products}
            ),
        },
        "tiers": tier_rows,
        "products": product_rows,
        "character_products": [
            item for item in character_products if item["character_id"] not in anonymous_character_ids
        ],
    }


def _aggregate_seed(row: PlanetaryProductionSnapshot) -> dict[str, Any]:
    return {
        "character_id": row.character_id,
        "character_name": row.character.name,
        "product_type_id": row.product_type_id,
        "product_name": row.product_type.name,
        "tier": row.commodity_tier,
        "estimated_units": 0.0,
        "estimated_volume": 0.0,
        "current_units_per_day": 0.0,
        "current_volume_per_day": 0.0,
    }


def _windowed_estimate(row: PlanetaryProductionSnapshot, cutoff: datetime) -> float:
    units = float(row.estimated_units_since_previous or 0)
    started = _aware(row.interval_started_at)
    captured = _aware(row.captured_at) or row.captured_at
    if not started or started >= cutoff or captured <= cutoff:
        return units
    total = (captured - started).total_seconds()
    visible = (captured - cutoff).total_seconds()
    return units * max(0, min(visible / total, 1)) if total > 0 else units


def _empty_summary(days: int) -> dict[str, Any]:
    return {
        "days": days,
        "has_history": False,
        "cards": {
            "estimated_volume": 0,
            "current_volume_per_day": 0,
            "product_count": 0,
            "character_count": 0,
        },
        "tiers": [
            {
                "tier": tier,
                "label": label,
                "estimated_units": 0,
                "estimated_volume": 0,
                "current_units_per_day": 0,
                "current_volume_per_day": 0,
                "product_count": 0,
                "character_count": 0,
            }
            for tier, label in PI_TIER_LABELS.items()
        ],
        "products": [],
        "character_products": [],
    }

File: rust/eqm-core/src/planetary_analytics.rs
use std::cmp::Ordering;
use std::collections::{HashMap, HashSet};

use chrono::{DateTime, Utc};
use serde::Deserialize;
use serde_json::{json, Value};

const SCHEMA: &str = "eqm.planetary-analytics.v1";
const TIERS: [(&str, &str); 5] = [
    ("P0", "Raw resources"),
    ("P1", "Basic commodities"),
    ("P2", "Refined commodities"),
    ("P3", "Specialized commodities"),
    ("P4", "Advanced commodities"),
];

#[derive(Debug, Deserialize)]
pub struct PlanetaryAnalyticsInput {
    pub schema_version: String,
    pub days: i64,
    pub cutoff: String,
    #[serde(default)]
    pub history: Vec<SnapshotRow>,
    #[serde(default)]
    pub current: Vec<SnapshotRow>,
    #[serde(default)]
    pub anonymous_character_ids: Vec<i64>,
}

#[derive(Debug, Deserialize)]
pub struct SnapshotRow {
    pub character_id: i64,
    pub character_name: String,
    pub product_type_id: i64,
    pub product_name: String,
    pub tier: String,
    pub unit_volume: f64,
    #[serde(default)]
    pub estimated_units_since_previous: f64,
    #[serde(default)]
    pub projected_units_per_day: f64,
    #[serde(default)]
    pub interval_started_at: Option<String>,
    pub captured_at: String,
}

#[derive(Debug, Clone)]
struct CharacterProduct {
    character_id: i64,
    character_name: String,
    product_type_id: i64,
    product_name: String,
    tier: String,
    estimated_units: f64,
    estimated_volume: f64,
    current_units_per_day: f64,
    current_volume_per_day: f64,
    order: usize,
}

fn parse_utc(value: &str, label: &str) -> Result<DateTime<Utc>, String> {
    DateTime::parse_from_rfc3339(value)
        .map(|stamp| stamp.with_timezone(&Utc))
        .map_err(|error| format!("invalid {label}: {error}"))
}

fn windowed_estimate(row: &SnapshotRow, cutoff: DateTime<Utc>) -> Result<f64, String> {
    let captured = parse_utc(&row.captured_at, "captured_at")?;
    let Some(started_at) = row.interval_started_at.as_deref() else {
        return Ok(row.estimated_units_since_previous);
    };
    let started = parse_utc(started_at, "interval_started_at")?;
    if started >= cutoff || captured <= cutoff {
        return Ok(row.estimated_units_since_previous);
    }
    let total = (captured - started).num_milliseconds() as f64 / 1_000.0;
    let visible = (captured - cutoff).num_milliseconds() as f64 / 1_000.0;
    if total > 0.0 {
        Ok(row.estimated_units_since_previous * (visible / total).clamp(0.0, 1.0))
    } else {
        Ok(row.estimated_units_since_previous)
    }
}

fn descending_pair(left: &CharacterProduct, right: &CharacterProduct) -> Ordering {
    right
        .estimated_volume
        .total_cmp(&left.estimated_volume)
        .then_with(|| {
            right
                .current_volume_per_day
                .total_cmp(&left.current_volume_per_day)
        })
        .then_with(|| left.order.cmp(&right.order))
}

fn character_product_json(row: &CharacterProduct) -> Value {
    json!({
        "character_id": row.character_id,
        "character_name": row.character_name,
        "product_type_id": row.product_type_id,
        "product_name": row.product_name,
        "tier": row.tier,
        "estimated_units": row.estimated_units,
        "estimated_volume": row.estimated_volume,
        "current_units_per_day": row.current_units_per_day,
        "current_volume_per_day": row.current_volume_per_day,
    })
}

fn ensure_character_product(
    rows: &mut Vec<CharacterProduct>,
    indexes: &mut HashMap<(i64, i64), usize>,
    source: &SnapshotRow,
) -> usize {
    let key = (source.character_id, source.product_type_id);
    if let Some(index) = indexes.get(&key) {
        return *index;
    }
    let index = rows.len();
    indexes.insert(key, index);
    rows.push(CharacterProduct {
        character_id: source.character_id,
        character_name: source.character_name.clone(),
        product_type_id: source.product_type_id,
        product_name: source.product_name.clone(),
        tier: source.tier.clone(),
        estimated_units: 0.0,
        estimated_volume: 0.0,
        current_units_per_day: 0.0,
        current_volume_per_day: 0.0,
        order: index,
    });
    index
}

pub fn evaluate_planetary_analytics(input: PlanetaryAnalyticsInput) -> Result<Value, String> {
    if input.schema_version != SCHEMA {
        return Err(format!(
            "unsupported planetary analytics schema: {}",
            input.schema_version
        ));
    }
    let cutoff = parse_utc(&input.cutoff, "cutoff")?;
    let anonymous: HashSet<i64> = input.anonymous_character_ids.into_iter().collect();
    let mut rows: Vec<CharacterProduct> = Vec::new();
    let mut indexes: HashMap<(i64, i64), usize> = HashMap::new();

    for source in &input.history {
        let units = windowed_estimate(source, cutoff)?;
        let index = ensure_character_product(&mut rows, &mut indexes, source);
        rows[index].estimated_units += units;
        rows[index].estimated_volume += units * source.unit_volume;
    }
    for source in &input.current {
        let index = ensure_character_product(&mut rows, &mut indexes, source);
        rows[index].current_units_per_day += source.projected_units_per_day;
        rows[index].current_volume_per_day += source.projected_units_per_day * source.unit_volume;
    }
    rows.sort_by(descending_pair);

    #[derive(Clone)]
    struct Product {
        product_type_id: i64,
        product_name: String,
        tier: String,
        estimated_units: f64,
        estimated_volume: f64,
        current_units_per_day: f64,
        current_volume_per_day: f64,
        top_character: Option<String>,
        top_score: f64,
        order: usize,
    }
    struct TierTotals {
        estimated_units: f64,
        estimated_volume: f64,
        current_units_per_day: f64,
        current_volume_per_day: f64,
        products: HashSet<i64>,
        characters: HashSet<i64>,
    }
    let mut products: Vec<Product> = Vec::new();
    let mut product_indexes: HashMap<i64, usize> = HashMap::new();
    let mut tier_totals: HashMap<String, TierTotals> = TIERS
        .iter()
        .map(|(tier, _)| {
            (
                (*tier).to_string(),
                TierTotals {
                    estimated_units: 0.0,
                    estimated_volume: 0.0,
                    current_units_per_day: 0.0,
                    current_volume_per_day: 0.0,
                    products: HashSet::new(),
                    characters: HashSet::new(),
                },
            )
        })
        .collect();

    for row in &rows {
        let product_index = *product_indexes
            .entry(row.product_type_id)
            .or_insert_with(|| {
                let index = products.len();
                products.push(Product {
                    product_type_id: row.product_type_id,
                    product_name: row.product_name.clone(),
                    tier: row.tier.clone(),
                    estimated_units: 0.0,
                    estimated_volume: 0.0,
                    current_units_per_day: 0.0,
                    current_volume_per_day: 0.0,
                    top_character: None,
                    top_score: -1.0,
                    order: index,
                });
                index
            });
        let product = &mut products[product_index];
        product.estimated_units += row.estimated_units;
        product.estimated_volume += row.estimated_volume;
        product.current_units_per_day += row.current_units_per_day;
        product.current_volume_per_day += row.current_volume_per_day;
        let score = if row.estimated_units != 0.0 {
            row.estimated_units
        } else {
            row.current_units_per_day
        };
        if !anonymous.contains(&row.character_id) && score > product.top_score {
            product.top_score = score;
            product.top_character = Some(row.character_name.clone());
        }
        let tier = tier_totals
            .get_mut(&row.tier)
            .ok_or_else(|| format!("unsupported commodity tier: {}", row.tier))?;
        tier.estimated_units += row.estimated_units;
        tier.estimated_volume += row.estimated_volume;
        tier.current_units_per_day += row.current_units_per_day;
        tier.current_volume_per_day += row.current_volume_per_day;
        tier.products.insert(row.product_type_id);
        tier.characters.insert(row.character_id);
    }
    products.sort_by(|left, right| {
        right
            .estimated_volume
            .total_cmp(&left.estimated_volume)
            .then_with(|| {
                right
                    .current_volume_per_day
                    .total_cmp(&left.current_volume_per_day)
            })
            .then_with(|| left.order.cmp(&right.order))
    });
    let product_rows: Vec<Value> = products
        .into_iter()
        .map(|row| {
            json!({
                "product_type_id": row.product_type_id,
                "product_name": row.product_name,
                "tier": row.tier,
                "estimated_units": row.estimated_units,
                "estimated_volume": row.estimated_volume,
                "current_units_per_day": row.current_units_per_day,
                "current_volume_per_day": row.current_volume_per_day,
                "top_character": row.top_character,
            })
        })
        .collect();
    let tier_rows: Vec<Value> = TIERS
        .iter()
        .map(|(tier, label)| {
            let totals = tier_totals.get(*tier).expect("tier initialized");
            json!({
                "tier": tier,
                "label": label,
                "estimated_units": totals.estimated_units,
                "estimated_volume": totals.estimated_volume,
                "current_units_per_day": totals.current_units_per_day,
                "current_volume_per_day": totals.current_volume_per_day,
                "product_count": totals.products.len(),
                "character_count": totals.characters.len(),
            })
        })
        .collect();
    let character_ids: HashSet<i64> = rows.iter().map(|row| row.character_id).collect();
    Ok(json!({
        "schema_version": SCHEMA,
        "days": input.days,
        "has_history": rows.iter().any(|row| row.estimated_units > 0.0),
        "cards": {
            "estimated_volume": rows.iter().map(|row| row.estimated_volume).sum::<f64>(),
            "current_volume_per_day": rows.iter().map(|row| row.current_volume_per_day).sum::<f64>(),
            "product_count": product_rows.len(),
            "character_count": character_ids.len(),
        },
        "tiers": tier_rows,
        "products": product_rows,
        "character_products": rows.iter().filter(|row| !anonymous.contains(&row.character_id)).map(character_product_json).collect::<Vec<_>>(),
    }))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn row(character_id: i64, product_type_id: i64, units: f64, rate: f64) -> SnapshotRow {
        SnapshotRow {
            character_id,
            character_name: format!("Pilot {character_id}"),
            product_type_id,
            product_name: format!("Product {product_type_id}"),
            tier: "P2".to_string(),
            unit_volume: 1.5,
            estimated_units_since_previous: units,
            projected_units_per_day: rate,
            interval_started_at: Some("2026-08-01T00:00:00+00:00".to_string()),
            captured_at: "2026-08-03T00:00:00+00:00".to_string(),
        }
    }

    #[test]
    fn aggregates_history_current_rates_and_anonymity() {
        let output = evaluate_planetary_analytics(PlanetaryAnalyticsInput {
            schema_version: SCHEMA.to_string(),
            days: 1,
            cutoff: "2026-08-02T00:00:00+00:00".to_string(),
            history: vec![row(1, 44, 100.0, 0.0), row(2, 44, 50.0, 0.0)],
            current: vec![row(1, 44, 0.0, 40.0), row(2, 44, 0.0, 20.0)],
            anonymous_character_ids: vec![1],
        })
        .unwrap();
        assert_eq!(output["cards"]["estimated_volume"], 112.5);
        assert_eq!(output["cards"]["current_volume_per_day"], 90.0);
        assert_eq!(output["products"][0]["top_character"], "Pilot 2");
        assert_eq!(output["character_products"].as_array().unwrap().len(), 1);
        assert_eq!(output["tiers"][2]["product_count"], 1);
    }
}
